src/net.cpp

```diff
@@ -449,102 +460,108 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
+                        }
                     }
                 }
-            }
 
-            if (connected) {
-                sock = std::move(conn.sock);
-                addr_bind = CAddress{conn.me, NODE_NONE};
+                if (connected) {
+                    sock = std::move(conn.sock);
+                    addr_bind = CAddress{conn.me, NODE_NONE};
+                }
+            } else if (use_proxy) {
+                sock = CreateSock(proxy.proxy);
+                if (!sock) {
+                    return nullptr;
+                }
+                connected = ConnectThroughProxy(proxy, target_addr.ToStringAddr(), target_addr.GetPort(),
+                                                *sock, nConnectTimeout, proxyConnectionFailed);
+            } else {
+                // no proxy needed (none set for target network)
+                sock = CreateSock(target_addr);
+                if (!sock) {
+                    continue;
+                }
+                connected = ConnectSocketDirectly(target_addr, *sock, nConnectTimeout,
+                                                conn_type == ConnectionType::MANUAL);
             }
-        } else if (use_proxy) {
-            sock = CreateSock(proxy.proxy);
-            if (!sock) {
-                return nullptr;
+            if (!proxyConnectionFailed) {
+                // If a connection to the node was attempted, and failure (if any) is not caused by a problem connecting to
+                // the proxy, mark this as an attempt.
+                addrman.Attempt(target_addr, fCountFailure);
             }
-            connected = ConnectThroughProxy(proxy, addrConnect.ToStringAddr(), addrConnect.GetPort(),
-                                            *sock, nConnectTimeout, proxyConnectionFailed);
-        } else {
-            // no proxy needed (none set for target network)
-            sock = CreateSock(addrConnect);
+        } else if (pszDest && GetNameProxy(proxy)) {
+            sock = CreateSock(proxy.proxy);
             if (!sock) {
                 return nullptr;
             }
-            connected = ConnectSocketDirectly(addrConnect, *sock, nConnectTimeout,
-                                              conn_type == ConnectionType::MANUAL);
+            std::string host;
+            uint16_t port{default_port};
+            SplitHostPort(std::string(pszDest), port, host);
+            bool proxyConnectionFailed;
+            connected = ConnectThroughProxy(proxy, host, port, *sock, nConnectTimeout,
+                                            proxyConnectionFailed);
         }
-        if (!proxyConnectionFailed) {
-            // If a connection to the node was attempted, and failure (if any) is not caused by a problem connecting to
-            // the proxy, mark this as an attempt.
-            addrman.Attempt(addrConnect, fCountFailure);
+
+        // Check any other resolved address (if any) if we fail to connect
+        if (!connected) {
```

Comment:
nit:

I'd suggest to clarify the commit message saying that "a valid one is found" refers to what exactly (getting up to here and `connected = true`).

Reply:
I changed that bit of the commit message to:

```
This patches the aforementioned behavior by going over all resolved IPs until we find one
we can connect to or until we exhaust them.
```

I think that should be clear enough given the variable name, but let me know otherwise